src/models/neural_networks/triplet_data.py

```diff
@@ -0,0 +1,235 @@
+        self.multiplier = multiplier
+
+    def __iter__(self):
+        return _TripletIter(self)
+
+    def __len__(self) -> int:
+        return len(self.data_files) // self.batch_file_size
+
+
+class _TripletIter(_BaseIter):
+    def __init__(self, loader: TripletLoader) -> None:
+        super().__init__(loader)
+        self.neighbouring_distance = loader.neighbouring_distance
+        self.multiplier = loader.multiplier
+
+    def __next__(
+        self,
+    ) -> Tuple[
+        Tuple[Optional[torch.Tensor], ...],
+        Tuple[Optional[torch.Tensor], ...],
+        Tuple[Optional[torch.Tensor], ...],
+    ]:
+
+        global_modelarrays: Optional[ModelArrays] = None
+
+        if self.idx < self.max_idx:
+
+            cur_max_idx = min(self.idx + self.batch_file_size, self.max_idx)
+            while self.idx < cur_max_idx:
+                subfolder = self.data_files[self.idx]
+                arrays = self.ds_folder_to_np(
+                    subfolder, clear_nans=self.clear_nans, to_tensor=False
+                )
+                if arrays.x.historical.shape[0] == 0:
+                    print(f"{subfolder} returns no values. Skipping")
+
+                    # remove the empty element from the list
+                    self.data_files.pop(self.idx)
+                    self.max_idx -= 1
+                    self.idx -= 1  # we're going to add one later
+
+                    cur_max_idx = min(cur_max_idx + 1, self.max_idx)
+
+                if global_modelarrays is None:
+                    global_modelarrays = arrays
+                else:
+                    global_modelarrays.concatenate(arrays)
+
+                self.idx += 1
+            if global_modelarrays is not None:
+
+                anchor, neighbour, distant = self.find_neighbours(
+                    global_modelarrays, self.neighbouring_distance, self.multiplier
+                )
+
+                if self.to_tensor:
+                    anchor.to_tensor(self.device)
+                    neighbour.to_tensor(self.device)
+                    distant.to_tensor(self.device)
+                    global_modelarrays.to_tensor(self.device)
+
+                return (
+                    anchor.return_tuple(),
+                    neighbour.return_tuple(),
+                    distant.return_tuple(),
+                )
+            else:
+                raise StopIteration()
+
+        else:  # final_x_curr >= self.max_idx
+            raise StopIteration()
+
+    @staticmethod
+    def find_neighbours(
+        x: ModelArrays, distance: Tuple[float, float], multiplier: int,
+    ) -> Tuple[TrainData, TrainData, TrainData]:
+        """
+        Given a single modelArray, returns a tuple containing
+        [AnchorData, NeighbourData, DistantData]
+        """
+        anchor_indices: List[int] = []
+        neighbour_indices: List[int] = []
+        distant_indices: List[int] = []
+
+        outer_distance = tuple(multiplier * val for val in distance)
```

Comment:
whats the role of the multiplier?

Reply:
It's basically to enforce a minimum distance between the neighbouring instance and the distant instance.

The neighbour will be within `neighbouring_distance` of the anchor. The distant instance will be further than `multiplier * neighbouring_distance` from the anchor

Reply:
gotcha! so basically enforcing how large an area our spatial differences should be over

Reply:
yup!